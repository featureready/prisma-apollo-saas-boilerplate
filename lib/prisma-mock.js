import bcrypt from 'bcrypt';
import _ from 'lodash';

const users = [
  {
    id: 1,
    firstName: 'Test',
    lastName: 'User',
    email: 'test+user@email.com',
    createdAt: new Date(1988, 2, 2),
    updatedAt: new Date(1988, 2, 3),
    periodStart: new Date(1988, 10, 10),
    periodEnd: new Date(1988, 10, 11),
    password: bcrypt.hashSync('testpassword', 10),
    resetPasswordToken: 'RESET-PASSWORD-TOKEN',
    team: () => ({
      id: 1,
      name: 'A team name'
    })
  },
  {
    id: 2,
    firstName: 'Test',
    lastName: 'User',
    email: 'user+no+team@email.com',
    createdAt: new Date(1988, 2, 2),
    updatedAt: new Date(1988, 2, 3),
    periodStart: new Date(1988, 10, 10),
    periodEnd: new Date(1988, 10, 11),
    password: bcrypt.hashSync('testpassword', 10),
    resetPasswordToken: 'RESET-PASSWORD-TOKEN',
    team: () => null
  }
];

export const prisma = {
  users: jest.fn(() => users),
  user: jest.fn(({ id, email, resetPasswordToken }) => {
    if (id) return _.find(users, { id });
    if (email) return _.find(users, { email });
    if (resetPasswordToken) return _.find(users, { resetPasswordToken });

    return null;
  }),
  team: jest.fn(() => ({
    id: 1,
    name: 'A team name'
  })),
  updateTeam: jest.fn(fields => ({
    ...prisma.team()
  })),
  createUser: jest.fn(fields => ({
    ..._.first(users),
    ...fields,
    password: bcrypt.hashSync(fields.password, 10)
  })),
  updateUser: jest.fn(fields => ({
    ..._.first(users),
    ...fields,
    team: prisma.team
  })),
  deleteUser: jest.fn()
};
export const test = {};
