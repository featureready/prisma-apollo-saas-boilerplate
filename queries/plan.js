import { gql } from 'apollo-server-express';

export const PLANS = gql`
  query {
    Plans {
      id
      amount
      currency
      interval
      title
      description
    }
  }
`;

export default {
  PLANS
};
