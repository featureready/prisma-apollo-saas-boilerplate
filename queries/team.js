import { gql } from 'apollo-server-express';

export const CREATE_TEAM = `
mutation CreateTeam($input: CreateTeamInput!) {
  CreateTeam(input: $input) {
    id
    name
  }
}
`;

export const UPDATE_TEAM = `
mutation UpdateTeam($input: UpdateTeamInput!) {
  UpdateTeam(input: $input) {
    id
    name
  }
}
`;
