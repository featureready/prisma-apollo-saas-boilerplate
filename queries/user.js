import { gql } from 'apollo-server-express';

export const ME = gql`
  query {
    Me {
      id
      firstName
      lastName
      fullName
      email
      createdAt
      updatedAt
      periodStart
      periodEnd
      team {
        id
        name
      }
    }
  }
`;

export const PAYMENT_HISTORY = `
query {
  PaymentHistory {
    amountDue
    amountPaid
    invoicePdf
    status
    date
    periodStart
    periodEnd
  }
}
`;

export const SIGNUP = `
mutation Signup($input: SignupInput!) {
  Signup(input: $input) {
    jwt
    user {
      id
      firstName
      lastName
      email
    }
  }
}
`;

export const LOGIN = `
mutation Login($input: LoginInput!) {
  Login(input: $input) {
    jwt
    user {
      id
      firstName
      lastName
      email
    }
  }
}
`;

export const UPDATE_USER = `
mutation UpdateUser($input: UpdateUserInput!) {
  UpdateUser(input: $input) {
    id
    firstName
    lastName
    email
  }
}
`;

export const FORGOT_PASSWORD = `
mutation ForgotPassword($input: ForgotPasswordInput!) {
  ForgotPassword(input: $input) {
    message
  }
}
`;

export const RESET_PASSWORD = `
mutation ResetPassword($input: ResetPasswordInput!) {
  ResetPassword(input: $input) {
    message
  }
}
`;

export const DELETE_USER = `
mutation DeleteUser {
  DeleteUser {
    id
  }
}
`;

export const ADD_CREDIT_CARD = `
mutation AddCreditCard($input: AddCreditCardInput!) {
  AddCreditCard(input: $input) {
    message
  }
}
`;

export const SUBSCRIBE_PLAN = `
mutation SubscribePlan($input: SubscribePlanInput!) {
  SubscribePlan(input: $input) {
    message
  }
}
`;

export const INVITE_USER = `
mutation InviteUser($input: InviteUserInput!) {
  InviteUser(input: $input) {
    email
    firstName
    lastName
    status
  }
}
`;
