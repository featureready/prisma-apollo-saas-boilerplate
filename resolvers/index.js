import { GraphQLDateTime } from 'graphql-iso-date';

import userResolvers from './user';
import teamResolvers from './team';

const customScalarResolver = {
  DateTime: GraphQLDateTime
};

export default [customScalarResolver, userResolvers, teamResolvers];
