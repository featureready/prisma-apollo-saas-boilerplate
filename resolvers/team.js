import { UserInputError } from 'apollo-server';

export default {
  Mutation: {
    async CreateTeam(
      parent,
      {
        input: { name }
      },
      { user, prisma },
      info
    ) {
      const team = await prisma.user({ id: user.id }).team();
      if (team) {
        throw new UserInputError('You have already created a team.');
      } else {
        return prisma
          .updateUser({
            where: { id: user.id },
            data: {
              team: {
                create: {
                  name
                }
              }
            }
          })
          .team();
      }
    },
    async UpdateTeam(
      parent,
      {
        input: { name }
      },
      { user, prisma },
      info
    ) {
      const team = await prisma.user({ id: user.id }).team();
      if (team) {
        return prisma.updateTeam({
          where: { id: team.id },
          data: { name }
        });
      }

      throw new UserInputError('You have not created team yet.');
    }
  }
};
