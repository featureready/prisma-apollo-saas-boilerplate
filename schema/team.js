import { gql } from 'apollo-server-express';

export default gql`
  extend type Mutation {
    CreateTeam(input: CreateTeamInput!): Team!
      @requireAuth
      @analytics(type: "track", event: "Create Team")
      @analytics_group(type: "group")
    UpdateTeam(input: UpdateTeamInput!): Team!
      @requireAuth
      @analytics(type: "track", event: "Update Team")
  }

  input CreateTeamInput {
    name: String!
  }

  input UpdateTeamInput {
    name: String!
  }

  type Team {
    id: String
    name: String
  }
`;
