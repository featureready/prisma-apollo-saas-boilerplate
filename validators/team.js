import * as yup from 'yup';

export default yup.object().shape({
  name: yup.string().min(1, 'Team name must be at least 1 character')
});
