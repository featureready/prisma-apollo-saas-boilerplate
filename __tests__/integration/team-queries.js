import { createTestClient } from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import sgMail from '@sendgrid/mail';
import Stripe from 'stripe';
import { prisma } from '../../lib/prisma-mock';
import typeDefs from '../../schema';
import resolvers from '../../resolvers';
import schemaDirectives from '../../directives';
import { CREATE_TEAM, UPDATE_TEAM } from '../../queries/team';

let server;
let client;

beforeEach(async () => {
  server = await new ApolloServer({
    typeDefs,
    resolvers,
    schemaDirectives,
    context: () => ({
      prisma
    })
  });
  client = createTestClient(server);
});

it('creating a team when team already exists returns an error', async () => {
  server.context = () => ({
    prisma,
    user: { id: 1, email: 'test+user@email.com' }
  });

  client = createTestClient(server);
  const res = await client.query({
    query: CREATE_TEAM,
    variables: {
      input: {
        name: 'Team name'
      }
    }
  });

  expect(prisma.user).toHaveBeenCalledWith({ id: 1 });
  expect(res).toMatchSnapshot();
});

it('able to create team for user', async () => {
  server.context = () => ({
    prisma,
    user: { id: 2, email: 'user+no+team@email.com' }
  });

  client = createTestClient(server);
  const res = await client.query({
    query: CREATE_TEAM,
    variables: {
      input: {
        name: 'A new team name'
      }
    }
  });

  expect(prisma.user).toHaveBeenCalledWith({ id: 2 });
  expect(prisma.updateUser).toHaveBeenCalledWith({
    where: { id: 2 },
    data: {
      team: {
        create: {
          name: 'A new team name'
        }
      }
    }
  });
  expect(res).toMatchSnapshot();
});

it('able to update team name for user', async () => {
  server.context = () => ({
    prisma,
    user: { id: 1, email: 'test+user@email.com' }
  });

  client = createTestClient(server);
  const res = await client.query({
    query: UPDATE_TEAM,
    variables: {
      input: {
        name: 'Updated team name'
      }
    }
  });

  expect(prisma.user).toHaveBeenCalledWith({ id: 1 });
  expect(prisma.updateTeam).toHaveBeenCalledWith({
    where: { id: 1 },
    data: {
      name: 'Updated team name'
    }
  });
  expect(res).toMatchSnapshot();
});
