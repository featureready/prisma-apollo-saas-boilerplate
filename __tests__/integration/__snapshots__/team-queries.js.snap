// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`able to create team for user 1`] = `
Object {
  "data": Object {
    "CreateTeam": Object {
      "id": "1",
      "name": "A team name",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`able to update team name for user 1`] = `
Object {
  "data": Object {
    "UpdateTeam": Object {
      "id": "1",
      "name": "A team name",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`creating a team when team already exists returns an error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You have already created a team.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
